pkg/controllers/deprovisioning/orchestration/queue.go

```diff
@@ -0,0 +1,325 @@
+	queue := &Queue{
+		RateLimitingInterface:          workqueue.NewRateLimitingQueue(workqueue.NewItemExponentialFailureRateLimiter(queueBaseDelay, queueMaxDelay)),
+		candidateProviderIDToCommandID: map[string]uint64{},
+		idToCommand:                    map[uint64]*Command{},
+		kubeClient:                     kubeClient,
+		recorder:                       recorder,
+		cluster:                        cluster,
+		clock:                          clock,
+	}
+	go queue.Start(logging.WithLogger(ctx, logging.FromContext(ctx).Named("deprovisioning")))
+	return queue
+}
+
+// NewCommand creates a command key and adds in initial data for the orchestration queue.
+func NewCommand(replacements []nodeclaim.Key, candidates []*state.StateNode, reason string, timeAdded time.Time) *Command {
+	return &Command{
+		ReplacementKeys: lo.Map(replacements, func(key nodeclaim.Key, _ int) *NodeClaimKey {
+			return &NodeClaimKey{Key: key}
+		}),
+		Candidates: candidates,
+		Reason:     reason,
+		TimeAdded:  timeAdded,
+	}
+}
+
+// Start will kick off the queue. This runs asynchronously unil Karpenter shuts down.
+func (q *Queue) Start(ctx context.Context) {
+	for {
+		// Get command from queue. This waits until queue is non-empty.
+		item, shutdown := q.RateLimitingInterface.Get()
+		if shutdown {
+			break
+		}
+		hash := item.(uint64)
+		cmd := q.idToCommand[hash]
+		requeue, err := q.Handle(ctx, cmd)
+		if !requeue {
+			// If the command timed out, log the last error.
+			if err != nil {
+				// If the command failed, bail on the action.
+				// 1. Emit metrics for launch failures
+				// 2. Ensure cluster state no longer thinks these nodes are deleting
+				// 3. Remove it from the Queue's internal data structure
+				deprovisioningReplacementNodeLaunchFailedCounter.WithLabelValues(cmd.Reason).Inc()
+				q.cluster.UnmarkForDeletion(lo.Map(cmd.Candidates, func(s *state.StateNode, _ int) string { return s.ProviderID() })...)
+				err = multierr.Append(err, q.setNodesUnschedulable(ctx, false, lo.Map(cmd.Candidates, func(s *state.StateNode, _ int) string { return s.Node.Name })...))
+				logging.FromContext(ctx).Debugf("handling deprovisioning command, %w, for nodes %s", err,
+					strings.Join(lo.Map(cmd.Candidates, func(s *state.StateNode, _ int) string {
+						return s.Name()
+					}), ","))
+			}
+			q.Remove(hash, cmd)
+			continue
+		}
+		q.RateLimitingInterface.Done(hash)
+		// Requeue command if not complete
+		q.RateLimitingInterface.AddRateLimited(hash)
+	}
+	logging.FromContext(ctx).Errorf("deprovisioning queue is broken and has shutdown")
+}
+
+// Add adds commands to the Queue
+// Each command added to the queue should already be validated and ready for execution.
+func (q *Queue) Add(cmd *Command) error {
+	providerIDs := lo.Map(cmd.Candidates, func(s *state.StateNode, _ int) string {
+		return s.ProviderID()
+	})
+	// First check if we can add the command.
+	if err := q.CanAdd(providerIDs...); err != nil {
```

Comment:
Should this ever happen? I'd be surprised if this happened since we have full control over this queue and we are making Adds() to this cool in a synchronous manner.

Reply:
I don't think it hurts to check. The complexity of this, especially as we lean to parallelize more of this makes me want to err on the side of checking this. Thoughts?